burn-dataset/src/dataset/sqlite.rs

```diff
@@ -131,39 +228,380 @@ fn fetch_columns_and_len(
+
+                home_dir.join(".cache").join("burn-dataset")
+            }
+        }
+    }
+
+    /// Provides a writer instance for the SQLite dataset.
+    ///
+    /// # Arguments
+    ///
+    /// * `overwrite` - A boolean indicating if the existing database file should be overwritten.
+    ///
+    /// # Returns
+    ///
+    /// * A `Result` which is `Ok` if the writer could be created, `Err` otherwise.
+    pub fn writer<I>(self, overwrite: bool) -> Result<SqliteDatasetWriter<I>>
+    where
+        I: Clone + Send + Sync + Serialize + DeserializeOwned,
+    {
+        SqliteDatasetWriter::new(self.db_file(), overwrite)
+    }
+
+    /// Provides a reader instance for the SQLite dataset.
+    ///
+    /// # Arguments
+    ///
+    /// * `split` - A string slice that defines the data split for reading (e.g., "train", "test").
+    ///
+    /// # Returns
+    ///
+    /// * A `Result` which is `Ok` if the reader could be created, `Err` otherwise.
+    pub fn reader<I>(self, split: &str) -> Result<SqliteDataset<I>>
+    where
+        I: Clone + Send + Sync + Serialize + DeserializeOwned,
+    {
+        if !self.exists() {
+            panic!("The database file does not exist");
+        }
+
+        SqliteDataset::from_db_file(self.db_file(), split)
+    }
+}
+
+/// This `SqliteDatasetWriter` struct is a SQLite database writer dedicated to storing datasets.
+/// It retains the current writer's state and its database connection.
+///
+/// Being thread-safe, this writer can be concurrently used across multiple threads.
+///
+/// Typical applications include:
+///
+/// - Generation of a new dataset
+/// - Storage of preprocessed data or metadata
+/// - Enlargement of a dataset's item count post preprocessing
+#[derive(Debug)]
+pub struct SqliteDatasetWriter<I> {
+    db_file: PathBuf,
+    db_file_tmp: Option<NamedTempFile>,
+    splits: Arc<RwLock<HashSet<String>>>,
+    overwrite: bool,
+    conn_pool: Option<Pool<SqliteConnectionManager>>,
+    is_completed: Arc<RwLock<bool>>,
+    phantom: PhantomData<I>,
+}
+
+impl<I> SqliteDatasetWriter<I>
+where
+    I: Clone + Send + Sync + Serialize + DeserializeOwned,
+{
+    /// Creates a new instance of `SqliteDatasetWriter`.
+    ///
+    /// # Arguments
+    ///
+    /// * `db_file` - A reference to the Path that represents the database file path.
+    /// * `overwrite` - A boolean indicating if the existing database file should be overwritten.
+    ///
+    /// # Returns
+    ///
+    /// * A `Result` which is `Ok` if the writer could be created, `Err` otherwise.
+    pub fn new<P: AsRef<Path>>(db_file: P, overwrite: bool) -> Result<Self> {
+        let writer = Self {
+            db_file: db_file.as_ref().to_path_buf(),
+            db_file_tmp: None,
+            splits: Arc::new(RwLock::new(HashSet::new())),
+            overwrite,
+            conn_pool: None,
+            is_completed: Arc::new(RwLock::new(false)),
```

Comment:
Could be an atomic bool instead.

Reply:
Yes, I have looked into using AtomicBool. Though AtomicBool is safe to use for (read bool and immediate update bool), it's non-blocking. I need to block writer's write if the "complete" procedure is in the process of IO changes. I only update is_completed to true at the end if all IO finishes successfully. Setting is_completed at the beginning will not be technically correct. As far as I understood, AtomicBool is mainly for race conditions at the instruction level pipeline, so that the read and updating is in one go. In our case, we are doing IO.